File: qc-core/src/main/java/com/eager/questioncloud/core/domain/creator/infrastructure/CreatorStatisticsRepositoryImpl.java
package com.eager.questioncloud.core.domain.creator.infrastructure;

import static com.eager.questioncloud.core.domain.creator.infrastructure.QCreatorStatisticsEntity.creatorStatisticsEntity;

import com.eager.questioncloud.core.domain.creator.model.CreatorStatistics;
import com.eager.questioncloud.core.exception.CoreException;
import com.eager.questioncloud.core.exception.Error;
import com.querydsl.jpa.impl.JPAQueryFactory;
import java.util.List;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

@Repository
@RequiredArgsConstructor
public class CreatorStatisticsRepositoryImpl implements CreatorStatisticsRepository {
    private final CreatorStatisticsJpaRepository creatorStatisticsJpaRepository;
    private final JPAQueryFactory jpaQueryFactory;

    @Override
    public void save(CreatorStatistics creatorStatistics) {
        creatorStatisticsJpaRepository.save(CreatorStatisticsEntity.from(creatorStatistics));
    }

    @Override
    public void saveAll(List<CreatorStatistics> creatorStatistics) {
        creatorStatisticsJpaRepository.saveAll(
            creatorStatistics.stream().map(CreatorStatisticsEntity::from).collect(Collectors.toList())
        );
    }

    @Override
    public CreatorStatistics findByCreatorId(Long creatorId) {
        return creatorStatisticsJpaRepository.findByCreatorId(creatorId)
            .orElseThrow(() -> new CoreException(Error.NOT_FOUND))
            .toModel();
    }

    @Override
    @Transactional
    public void addSalesCount(Long creatorId, int count) {
        jpaQueryFactory.update(creatorStatisticsEntity)
            .set(creatorStatisticsEntity.salesCount, creatorStatisticsEntity.salesCount.add(count))
            .where(creatorStatisticsEntity.creatorId.eq(creatorId))
            .execute();
    }

    @Override
    public void deleteAllInBatch() {
        creatorStatisticsJpaRepository.deleteAllInBatch();
    }
}

File: qc-core/qc-creator/qc-creator-rdb/src/main/kotlin/com/eager/questioncloud/creator/repository/CreatorStatisticsRepositoryImpl.kt
package com.eager.questioncloud.creator.repository

import com.eager.questioncloud.common.exception.CoreException
import com.eager.questioncloud.common.exception.Error
import com.eager.questioncloud.creator.domain.CreatorStatistics
import com.eager.questioncloud.creator.entity.CreatorStatisticsEntity
import com.eager.questioncloud.creator.entity.QCreatorStatisticsEntity.creatorStatisticsEntity
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

@Repository
class CreatorStatisticsRepositoryImpl(
    private val creatorStatisticsJpaRepository: CreatorStatisticsJpaRepository,
    private val jpaQueryFactory: JPAQueryFactory,
) : CreatorStatisticsRepository {
    override fun save(creatorStatistics: CreatorStatistics) {
        creatorStatisticsJpaRepository.save(CreatorStatisticsEntity.createNewEntity(creatorStatistics))
    }
    
    override fun update(creatorStatistics: CreatorStatistics) {
        creatorStatisticsJpaRepository.save(CreatorStatisticsEntity.fromExisting(creatorStatistics))
    }
    
    override fun findByCreatorId(creatorId: Long): CreatorStatistics {
        return creatorStatisticsJpaRepository.findByCreatorId(creatorId)
            .orElseThrow { CoreException(Error.NOT_FOUND) }
            .toModel()
    }
    
    override fun getForUpdate(creatorId: Long): CreatorStatistics {
        return creatorStatisticsJpaRepository.getForUpdate(creatorId)?.toModel() ?: throw CoreException(Error.NOT_FOUND)
    }
    
    override fun findByCreatorIdIn(creatorIds: List<Long>): Map<Long, CreatorStatistics> {
        return creatorStatisticsJpaRepository.findByCreatorIdIn(creatorIds)
            .stream()
            .map { entity -> entity.toModel() }
            .toList()
            .associateBy { it.creatorId }
    }
    
    @Transactional
    override fun addSalesCount(creatorId: Long, count: Int) {
        jpaQueryFactory.update(creatorStatisticsEntity)
            .set(
                creatorStatisticsEntity.salesCount,
                creatorStatisticsEntity.salesCount.add(count)
            )
            .where(creatorStatisticsEntity.creatorId.eq(creatorId))
            .execute()
    }
    
    @Transactional
    override fun incrementSubscribeCount(creatorId: Long) {
        jpaQueryFactory.update(creatorStatisticsEntity)
            .set(creatorStatisticsEntity.subscriberCount, creatorStatisticsEntity.subscriberCount.add(1))
            .where(creatorStatisticsEntity.creatorId.eq(creatorId))
            .execute()
    }
    
    @Transactional
    override fun decrementSubscribeCount(creatorId: Long) {
        jpaQueryFactory.update(creatorStatisticsEntity)
            .set(creatorStatisticsEntity.subscriberCount, creatorStatisticsEntity.subscriberCount.subtract(1))
            .where(creatorStatisticsEntity.creatorId.eq(creatorId))
            .execute()
    }
    
    override fun deleteAllInBatch() {
        creatorStatisticsJpaRepository.deleteAllInBatch()
    }
}
